File: src/planner/pkg_nav/src/test/local_path_server.cpp
#include "ros/ros.h"
#include "actionlib/server/simple_action_server.h"
#include "pkg_nav/local_pathAction.h"

typedef actionlib::SimpleActionServer<pkg_nav::local_pathAction> Server;

void cb(const pkg_nav::local_pathGoalConstPtr &goal, Server *server)
{
    //获取目标值
    int num = goal->num;
    ROS_INFO("目标值:%d", num);
    //累加并响应连续反馈
    int result = 0;
    pkg_nav::local_pathFeedback feedback; //连续反馈
    ros::Rate rate(1);                    //通过频率设置休眠时间
    for (int i = 1; i <= num; i++)
    {
        result += i;
        //组织连续数据并发布
        feedback.progress_bar = i / (double)num;
        ROS_INFO("最终结果:%.2f", 1.0 * i / num);
        server->publishFeedback(feedback);
        rate.sleep();
    }
    //设置最终结果
    pkg_nav::local_pathResult r;
    r.result = result;
    server->setSucceeded(r);
    ROS_INFO("最终结果:%d", r.result);
}

int main(int argc, char *argv[])
{
    setlocale(LC_ALL, "");
    ROS_INFO("action服务端实现");
    // 2.初始化ROS节点;
    ros::init(argc, argv, "local_path_server");
    // 3.创建NodeHandle;
    ros::NodeHandle nh;
    // 4.创建action服务对象;
    /*SimpleActionServer(ros::NodeHandle n, 
                        std::string name, 
                        boost::function<void (const pkg_nav::local_pathGoalConstPtr &)> execute_callback, 
                        bool auto_start)
    */
    // actionlib::SimpleActionServer<pkg_nav::local_pathAction> server(....);
    Server server(nh, "local_path", boost::bind(&cb, _1, &server), false);
    server.start();
    // 5.处理请求,产生反馈与响应;

    // 6.spin().
    ros::spin();
    return 0;
}

File: src/planner/pkg_nav/scripts/local_path_server.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
import actionlib
from pkg_nav.msg import *
from nav_msgs.msg import Path, OccupancyGrid,Odometry

"""
    需求:
        创建两个ROS 节点，服务器和客户端，
        客户端可以向服务器发送目标数据N(一个整型数据)服务器会计算 1 到 N 之间所有整数的和,
        这是一个循环累加的过程，返回给客户端，这是基于请求响应模式的，
        又已知服务器从接收到请求到产生响应是一个耗时操作，每累加一次耗时0.1s，
        为了良好的用户体验，需要服务器在计算过程中，
        每累加一次，就给客户端响应一次百分比格式的执行进度，使用 action实现。
    流程:
        1.导包
        2.初始化 ROS 节点
        3.使用类封装，然后创建对象
        4.创建服务器对象
        5.处理请求数据产生响应结果，中间还要连续反馈
        6.spin
"""

class MyActionServer:
    def __init__(self):
        #SimpleActionServer(name, ActionSpec, execute_cb=None, auto_start=True)
        self.server = actionlib.SimpleActionServer("/ow/local_path",local_pathAction,self.cb,False)
        self.server.start() 
        self.local_path = Path()
        rospy.loginfo("The local_path 服务端启动")


    def cb(self,goal):
        rospy.loginfo("The local_path 服务端处理请求:")
        #1.解析目标值
        num = goal.num
        # self.local_path = goal.local_path
        #2.循环累加，连续反馈
        rate = rospy.Rate(1)
        sum = 0
        for i in range(1,num + 1):
            # 累加
            sum = sum + i
            # 计算进度并连续反馈
            feedBack = i*1.0 / num
            rospy.loginfo("当前进度:%.2f/100 ",feedBack*100.0)
            rospy.loginfo("goal.runCommand = %d ",goal.runCommand)
            rospy.loginfo("goal.local_path.size =  ")
            # print(self.local_path.poses[i].pose.position)
            print(goal.local_path.poses[0].pose.position)
            # if i == 8:
              # self.server.set_aborted()

            feedBack_obj = local_pathFeedback()
            feedBack_obj.progress_bar = feedBack
            self.server.publish_feedback(feedBack_obj)
            rate.sleep()
        #3.响应最终结果
        result = local_pathResult()
        result.result = sum        
        self.server.set_succeeded(result)
        # self.server.set_aborted()
        rospy.loginfo("响应结果:%d ",sum)
        
if __name__ == "__main__":
    rospy.init_node("action_server_p")
    server = MyActionServer()
    rospy.spin()